File: streamlit_project/score_tracker.py
import streamlit as st
from influxdb import InfluxDBClient
from datetime import datetime
from urllib.parse import quote, unquote
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

def show_page(sunset_data, all_data, ranked_images):
    st.title("Score Tracker")
    st.write("Track and visualize sunset scores over time")
    
    # Convert to DataFrame and clean data
    data = pd.DataFrame(ranked_images)
    data["Score"] = pd.to_numeric(data["Score"], errors='coerce')
    
    # Convert Date to datetime for proper sorting
    data["dt_local"] = pd.to_datetime(data["dt_local"])
    
    # Extract separate date and time columns for axis options
    data["Date_Only"] = data["Date"]
    data["Time_Only"] = data["Time"]
    
    # Axis selection controls at the top
    col1, col2 = st.columns(2)
    with col1:
        x_axis = st.selectbox("X-Axis:", ["Date Only", "Time Only", "Score"], index=0)
    with col2:
        y_axis = st.selectbox("Y-Axis:", ["Score", "Date Only", "Time Only"], index=0)
    
    st.divider()
         
    if not data.empty:
        # Map axis selections to actual column names
        axis_mapping = {
            "Date Only": "Date_Only", 
            "Time Only": "Time_Only",
            "Score": "Score"
        }
        
        x_col = axis_mapping[x_axis]
        y_col = axis_mapping[y_axis]
        
        # Sort data by X-axis column to prevent crossing lines
        data_sorted = data.sort_values(by=x_col)
        
        # Create tabs for chart and data
        tab1, tab2 = st.tabs(["Line Chart","Data"])
        
        with tab1:
            # Create Plotly figure with click events
            fig = go.Figure()
            
            fig.add_trace(go.Scatter(
                x=data_sorted[x_col],
                y=data_sorted[y_col],
                mode='lines+markers',
                name='Sunset Data',
                customdata=data_sorted.index,  # Store row indices from sorted data
                hovertemplate="<b>%{text}</b><br>" +
                             f"{x_axis}: %{{x}}<br>" +
                             f"{y_axis}: %{{y:.2f}}<br>" +
                             "Click to see image<br>" +
                             "<extra></extra>",
                text=data_sorted["Label"],
                line=dict(width=2),
                marker=dict(size=8)
            ))
            
            fig.update_layout(
                title=f"{y_axis} vs {x_axis} (Click points to see images)",
                xaxis_title=x_axis,
                yaxis_title=y_axis,
                height=600,
                hovermode='closest'
            )
            
            # Add grid lines
            fig.update_xaxes(
                showgrid=True,
                gridwidth=1,
                gridcolor='rgba(128, 128, 128, 0.3)'
            )
            fig.update_yaxes(
                showgrid=True,
                gridwidth=1,
                gridcolor='rgba(128, 128, 128, 0.3)'
            )
            
            # Display chart and capture click events
            clicked_data = st.plotly_chart(fig, use_container_width=True, on_select="rerun")
            
            # Show images when point is clicked
            if clicked_data and clicked_data["selection"]["points"]:
                point_index = clicked_data["selection"]["points"][0]["customdata"]
                selected_row = data.iloc[point_index]  # Use original data index
                
                # Score info on left, images on right
                col1, col2 = st.columns([1, 3])
                
                with col1:
                    st.metric("Selected Score", f"{selected_row['Score']:.2f}")
                    st.write(f"**Label:** {selected_row['Label']}")
                    st.write(f"**Date:** {selected_row['dt_local'].strftime('%Y-%m-%d %H:%M')}")
                    st.write(f"**X-Value:** {selected_row[x_col]}")
                    st.write(f"**Y-Value:** {selected_row[y_col]}")
                
                with col2:
                    # Show both images side by side in right column
                    img_col1, img_col2 = st.columns(2)
                    with img_col1:
                        st.image(selected_row["Ranked Image"], caption="Ranked Image (with analysis)", width=400)
                    
                    with img_col2:
                        if "Raw Image" in selected_row and pd.notna(selected_row["Raw Image"]):
                            st.image(selected_row["Raw Image"], caption="Original Image", width=400)
                        else:
                            st.info("No original image available")
        
        with tab2:
            st.dataframe(
                data[["dt_local", "Date_Only", "Time_Only", "Score", "Label"]].sort_values("dt_local", ascending=False),
                height=400, 
                use_container_width=True
            )
        
        # Summary stats
        st.subheader("Summary Statistics")
        col1, col2, col3, col4 = st.columns(4)
        
        with col1:
            st.metric("Average Score", f"{data['Score'].mean():.1f}")
        with col2:
            st.metric("Highest Score", f"{data['Score'].max():.1f}")
        with col3:
            st.metric("Lowest Score", f"{data['Score'].min():.1f}")
        with col4:
            st.metric("Total Images", len(data))
            
    else:
        st.info("No data available.")

File: html_project/static/pages/scoretracker.js
// Mirrors streamlit_project/score_tracker.py

const AXIS_KEYS = {
  "Date Only": "Date",
  "Time Only": "Time",
  Score: "Score",
};

export function renderScoreTracker(container, state) {
  if (!state.scoretracker) {
    state.scoretracker = { xAxis: "Date Only", yAxis: "Score", tab: "chart", selectedIndex: null };
  }
  const st = state.scoretracker;

  container.innerHTML = "";

  const title = document.createElement("h2");
  title.textContent = "Score Tracker";
  container.appendChild(title);
  const subtitle = document.createElement("p");
  subtitle.textContent = "Track and visualize sunset scores over time";
  container.appendChild(subtitle);

  const rows = state.rankedImages;
  if (!rows.length) {
    const empty = document.createElement("div");
    empty.className = "empty";
    empty.textContent = "No data available.";
    container.appendChild(empty);
    return;
  }

  const controls = document.createElement("div");
  controls.className = "axis-controls";
  controls.append(
    makeSelect("X-Axis:", Object.keys(AXIS_KEYS), st.xAxis, (v) => {
      st.xAxis = v;
      renderScoreTracker(container, state);
    }),
    makeSelect("Y-Axis:", Object.keys(AXIS_KEYS), st.yAxis, (v) => {
      st.yAxis = v;
      renderScoreTracker(container, state);
    })
  );
  container.appendChild(controls);
  container.appendChild(document.createElement("hr"));

  const tabs = document.createElement("div");
  tabs.className = "tabs";
  for (const tabName of ["Line Chart", "Data"]) {
    const key = tabName === "Line Chart" ? "chart" : "data";
    const btn = document.createElement("button");
    btn.className = "tab-btn" + (st.tab === key ? " active" : "");
    btn.textContent = tabName;
    btn.addEventListener("click", () => {
      st.tab = key;
      renderScoreTracker(container, state);
    });
    tabs.appendChild(btn);
  }
  container.appendChild(tabs);

  const xCol = AXIS_KEYS[st.xAxis];
  const yCol = AXIS_KEYS[st.yAxis];

  const indexed = rows.map((r, i) => ({ ...r, _idx: i }));
  const sorted = [...indexed].sort((a, b) => (a[xCol] < b[xCol] ? -1 : a[xCol] > b[xCol] ? 1 : 0));

  const tabPane = document.createElement("div");
  container.appendChild(tabPane);

  if (st.tab === "chart") {
    renderChartTab(tabPane, state, sorted, xCol, yCol, st);
  } else {
    renderDataTab(tabPane, rows);
  }

  renderSummary(container, rows);
}

function renderChartTab(pane, state, sorted, xCol, yCol, st) {
  const chartDiv = document.createElement("div");
  chartDiv.id = "score-chart";
  pane.appendChild(chartDiv);

  const trace = {
    x: sorted.map((r) => r[xCol]),
    y: sorted.map((r) => r[yCol]),
    text: sorted.map((r) => r.Label),
    customdata: sorted.map((r) => r._idx),
    mode: "lines+markers",
    type: "scatter",
    name: "Sunset Data",
    hovertemplate: `<b>%{text}</b><br>${st.xAxis}: %{x}<br>${st.yAxis}: %{y:.2f}<br>Click to see image<extra></extra>`,
    line: { width: 2 },
    marker: { size: 8 },
  };

  const layout = {
    title: `${st.yAxis} vs ${st.xAxis} (Click points to see images)`,
    xaxis: { title: st.xAxis, showgrid: true, gridcolor: "rgba(128,128,128,0.3)" },
    yaxis: { title: st.yAxis, showgrid: true, gridcolor: "rgba(128,128,128,0.3)" },
    height: 600,
    hovermode: "closest",
    paper_bgcolor: "rgba(0,0,0,0)",
    plot_bgcolor: "rgba(0,0,0,0)",
    font: { color: "#f0f2f6" },
  };

  Plotly.newPlot(chartDiv, [trace], layout, { responsive: true });

  chartDiv.on("plotly_click", (evt) => {
    const point = evt.points[0];
    st.selectedIndex = point.customdata;
    renderPointDetail(pane, state, st.selectedIndex);
  });

  if (st.selectedIndex !== null) {
    renderPointDetail(pane, state, st.selectedIndex);
  }
}

function renderPointDetail(pane, state, index) {
  const existing = pane.querySelector(".point-detail");
  if (existing) existing.remove();

  const row = state.rankedImages[index];
  const wrap = document.createElement("div");
  wrap.className = "point-detail";

  const info = document.createElement("div");
  info.innerHTML = `
    <div class="metric-label">Selected Score</div>
    <div class="metric-value">${row.Score.toFixed(2)}</div>
    <div><strong>Label:</strong> ${row.Label}</div>
    <div><strong>Date:</strong> ${new Date(row.dt_local).toLocaleString()}</div>
  `;

  const images = document.createElement("div");
  images.className = "images";

  const rankedFig = document.createElement("figure");
  rankedFig.innerHTML = `<img src="${row["Ranked Image"]}" loading="lazy" /><figcaption>Ranked Image (with analysis)</figcaption>`;
  images.appendChild(rankedFig);

  const rawFig = document.createElement("figure");
  if (row["Raw Image"]) {
    rawFig.innerHTML = `<img src="${row["Raw Image"]}" loading="lazy" /><figcaption>Original Image</figcaption>`;
  } else {
    rawFig.innerHTML = `<div class="empty">No original image available</div>`;
  }
  images.appendChild(rawFig);

  wrap.append(info, images);
  pane.appendChild(wrap);
}

function renderDataTab(pane, rows) {
  const sorted = [...rows].sort((a, b) => new Date(b.dt_local) - new Date(a.dt_local));

  const table = document.createElement("table");
  table.innerHTML = `
    <thead>
      <tr><th>dt_local</th><th>Date</th><th>Time</th><th>Score</th><th>Label</th></tr>
    </thead>
    <tbody>
      ${sorted
        .map(
          (r) => `<tr>
            <td>${new Date(r.dt_local).toLocaleString()}</td>
            <td>${r.Date}</td>
            <td>${r.Time}</td>
            <td>${r.Score.toFixed(2)}</td>
            <td>${r.Label}</td>
          </tr>`
        )
        .join("")}
    </tbody>
  `;
  pane.appendChild(table);
}

function renderSummary(container, rows) {
  const scores = rows.map((r) => r.Score);
  const avg = scores.reduce((a, b) => a + b, 0) / scores.length;
  const max = Math.max(...scores);
  const min = Math.min(...scores);

  const heading = document.createElement("h4");
  heading.textContent = "Summary Statistics";
  container.appendChild(heading);

  const statsRow = document.createElement("div");
  statsRow.className = "stats-row";
  statsRow.append(
    stat("Average Score", avg.toFixed(1)),
    stat("Highest Score", max.toFixed(1)),
    stat("Lowest Score", min.toFixed(1)),
    stat("Total Images", rows.length)
  );
  container.appendChild(statsRow);
}

function stat(label, value) {
  const el = document.createElement("div");
  el.className = "stat";
  el.innerHTML = `<span class="metric-label">${label}</span><span class="metric-value">${value}</span>`;
  return el;
}

function makeSelect(labelText, options, value, onChange) {
  const wrap = document.createElement("div");
  const label = document.createElement("label");
  label.textContent = labelText;
  label.style.display = "block";
  label.style.marginBottom = "4px";
  label.style.color = "var(--muted)";
  label.style.fontSize = "0.85em";

  const select = document.createElement("select");
  for (const opt of options) {
    const o = document.createElement("option");
    o.value = opt;
    o.textContent = opt;
    if (opt === value) o.selected = true;
    select.appendChild(o);
  }
  select.addEventListener("change", () => onChange(select.value));

  wrap.append(label, select);
  return wrap;
}
